backends/aws_s3.py

```diff
@@ -0,0 +1,125 @@
+from functools import wraps
+from os import environ
+from os import getenv
+
+import boto3
+from botocore.exceptions import ClientError
+
+from backends.exceptions import ErrorException
+
+_ENV_ACCESS_KEY_ID = 'TP_BACKEND_AWS_S3_ACCESS_KEY_ID'
+_ENV_SECRET_ACCESS_KEY = 'TP_BACKEND_AWS_S3_SECRET_ACCESS_KEY'
+_ENV_BUCKET = 'TP_BACKEND_AWS_S3_BUCKET'
+
+_DEFAULT_BUCKET = 'torpaste'
+
+_s3 = None
+_bucket = None
+
+
+def _wrap_aws_exception(func):
+    @wraps(func)
+    def _adapt_exception_types(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except ClientError as ex:
+            raise ErrorException(
+                'Error while communicating with AWS S3') from ex
+
+    return _adapt_exception_types
+
+
+def _getenv_required(key):
+    try:
+        return environ[key]
+    except KeyError:
+        raise ErrorException(
+            'Required environment variable %s not set' % key)
+
+
+@_wrap_aws_exception
+def initialize_backend():
+    global _s3
+    global _bucket
+
+    _s3 = boto3.resource(
+        's3',
+        aws_access_key_id=_getenv_required(_ENV_ACCESS_KEY_ID),
+        aws_secret_access_key=_getenv_required(_ENV_SECRET_ACCESS_KEY))
+
+    _bucket = getenv(_ENV_BUCKET, _DEFAULT_BUCKET)
+    _s3.create_bucket(Bucket=_bucket)
+
+
+@_wrap_aws_exception
+def new_paste(paste_id, paste_content):
+    _s3.Bucket(_bucket).put_object(
+        Body=paste_content.encode('utf-8'),
+        Key=paste_id)
+
+
+@_wrap_aws_exception
+def update_paste_metadata(paste_id, metadata):
+    obj = _s3.Object(_bucket, paste_id)
+    obj.metadata.clear()
+    obj.metadata.update(metadata)
+    obj.copy_from(CopySource={'Bucket': _bucket, 'Key': paste_id},
+                  Metadata=obj.metadata,
+                  MetadataDirective='REPLACE')
+
+
+@_wrap_aws_exception
+def does_paste_exist(paste_id):
+    try:
+        _s3.Object(_bucket, paste_id).load()
```

Comment:
I assume the `load()` function will actually download the paste content. If there a different function to check whether a specific key exists? If not, `load()` is fine, but may cause double or triple downloads of the paste from the server.

Reply:
This should be fine. The documentation isn't entirely clear, but [Object.load](http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Object.load) should just do a minimal (HEAD) request to check if the object exists. [Object.get](http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Object.get), on the other hand, will do a full (GET) request to download the object's contents.